### PR TITLE
Residual kernels for the full viscous stress formulation

### PR DESCRIPTION
**Pull request by timofeymukha**
Adds the types to compute equation residuals for the full viscous stress formulation.

### CODE REVIEW

File: src/fluid/stress_formulation/bcknd/device/cuda/prs_stress_res_kernel.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+                                            const T * __restrict__ B,
+                                            T * __restrict__ h1,
+                                            const T * __restrict__ rho,
+                                            const int n) {
+  
+  const int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  const int str = blockDim.x * gridDim.x;
+  
+  for (int i = idx; i < n; i += str) {
+    //    h1[i] = inv_rho;
```
Comment (njansson) [inline src/fluid/stress_formulation/bcknd/device/cuda/prs_stress_res_kernel.h:57]:
```suggestion
```
Left over?

File: src/fluid/stress_formulation/bcknd/device/cuda/prs_stress_res_kernel.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+                                            T * __restrict__ ta2,
+                                            T * __restrict__ ta3,
+                                            const T * __restrict__ wa1,
+                                            const T * __restrict__ wa2,
+                                            const T * __restrict__ wa3,
+                                            const T * __restrict__ f_u,
+                                            const T * __restrict__ f_v,
+                                            const T * __restrict__ f_w,
+                                            const T * __restrict__ B,
+                                            T * __restrict__ h1,
```
Comment (njansson) [inline src/fluid/stress_formulation/bcknd/device/cuda/prs_stress_res_kernel.h:49]: If we don't use h1, we should not pass it as an argument.

```suggestion
```

File: src/fluid/stress_formulation/bcknd/device/cuda/pnpn_stress_res.cu [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    const dim3 nblcks(((*n) + 1024 - 1) / 1024, 1, 1);
+    const cudaStream_t stream = (cudaStream_t) glb_cmd_queue;
+    
+    prs_stress_res_part1_kernel<real>
+      <<<nblcks, nthrds, 0, stream>>>((real *) ta1, (real *) ta2,
+                                      (real *) ta3, (real *) wa1,
+                                      (real *) wa2, (real *) wa3,
+                                      (real *) f_u, (real *) f_v,
+                                      (real *) f_w, (real *) B,
+                                      (real *) h1,  (real *) rho, *n);
```
Comment (njansson) [inline src/fluid/stress_formulation/bcknd/device/cuda/pnpn_stress_res.cu:58]:
```suggestion
                                      (real *) rho, *n);
```
See comments in header file, if we don't use `h1` we shouldn't pass it.

**Review by njansson — COMMENTED** [review]
Looks good! There're some place where we should fuse some kernels to avoid launch latencies, but that's for another PR

File: src/fluid/stress_formulation/bcknd/device/pnpn_res_stress_device.F90 [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+#ifdef HAVE_HIP
+    call pnpn_prs_stress_res_part3_hip(p_res%x_d, ta1%x_d, ta2%x_d, ta3%x_d,  &
+                                       wa1%x_d, wa2%x_d, wa3%x_d, dtbd, n)
+#elif HAVE_CUDA
+    call pnpn_prs_stress_res_part3_cuda(p_res%x_d, ta1%x_d, ta2%x_d, ta3%x_d, &
+                                        wa1%x_d, wa2%x_d, wa3%x_d, dtbd, n)
+#elif HAVE_OPENCL
+    call pnpn_prs_stress_res_part3_opencl(p_res%x_d, ta1%x_d, ta2%x_d, ta3%x_d, &
+                                          wa1%x_d, wa2%x_d, wa3%x_d, dtbd, n)
+#endif
```
Comment (njansson) [inline src/fluid/stress_formulation/bcknd/device/pnpn_res_stress_device.F90:356]:
```suggestion
#ifdef HAVE_HIP
!    call pnpn_prs_stress_res_part3_hip(p_res%x_d, ta1%x_d, ta2%x_d, ta3%x_d,  &
!                                       wa1%x_d, wa2%x_d, wa3%x_d, dtbd, n)
#elif HAVE_CUDA
    call pnpn_prs_stress_res_part3_cuda(p_res%x_d, ta1%x_d, ta2%x_d, ta3%x_d, &
                                        wa1%x_d, wa2%x_d, wa3%x_d, dtbd, n)
#elif HAVE_OPENCL
!    call pnpn_prs_stress_res_part3_opencl(p_res%x_d, ta1%x_d, ta2%x_d, ta3%x_d, &
!                                          wa1%x_d, wa2%x_d, wa3%x_d, dtbd, n)
#endif
```

File: src/fluid/stress_formulation/bcknd/device/pnpn_res_stress_device.F90 [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    ! The laplacian of the pressure
+    call Ax%compute(p_res%x, p%x, c_Xh, p%msh, p%Xh)
+
+#ifdef HAVE_HIP
+    call pnpn_prs_res_part2_hip(p_res%x_d, wa1%x_d, wa2%x_d, wa3%x_d, n);
+#elif HAVE_CUDA
+    call pnpn_prs_res_part2_cuda(p_res%x_d, wa1%x_d, wa2%x_d, wa3%x_d, n);
+#elif HAVE_OPENCL
+    call pnpn_prs_res_part2_opencl(p_res%x_d, wa1%x_d, wa2%x_d, wa3%x_d, n);
+#endif
```
Comment (njansson) [inline src/fluid/stress_formulation/bcknd/device/pnpn_res_stress_device.F90:327]:
```suggestion
#ifdef HAVE_HIP
!    call pnpn_prs_res_part2_hip(p_res%x_d, wa1%x_d, wa2%x_d, wa3%x_d, n);
#elif HAVE_CUDA
    call pnpn_prs_res_part2_cuda(p_res%x_d, wa1%x_d, wa2%x_d, wa3%x_d, n);
#elif HAVE_OPENCL
 !   call pnpn_prs_res_part2_opencl(p_res%x_d, wa1%x_d, wa2%x_d, wa3%x_d, n);
#endif
```

File: src/fluid/stress_formulation/bcknd/device/pnpn_res_stress_device.F90 [hunk 6]
```diff
@@ -0,0 +1,416 @@
...
+
+#elif HAVE_CUDA
+    call pnpn_prs_stress_res_part1_cuda(ta1%x_d, ta2%x_d, ta3%x_d, &
+         wa1%x_d, wa2%x_d, wa3%x_d, f_x%x_d, f_y%x_d, f_z%x_d, &
+         c_Xh%B_d, c_Xh%h1_d, rho%x_d, n)
+#elif HAVE_OPENCL
+    call pnpn_prs_stress_res_part1_opencl(ta1%x_d, ta2%x_d, ta3%x_d, &
+         wa1%x_d, wa2%x_d, wa3%x_d, f_x%x_d, f_z%x_d, f_z%x_d, &
+         c_Xh%B_d, c_Xh%h1_d, rho%x_d, n)
+#endif
```
Comment (njansson) [inline src/fluid/stress_formulation/bcknd/device/pnpn_res_stress_device.F90:305]:
```suggestion
#ifdef HAVE_HIP
!    call pnpn_prs_stress_res_part1_hip(ta1%x_d, ta2%x_d, ta3%x_d, &
!         wa1%x_d, wa2%x_d, wa3%x_d, f_x%x_d, f_y%x_d, f_z%x_d, &
!         c_Xh%B_d, c_Xh%h1_d, rho%x_d, n)

#elif HAVE_CUDA
    call pnpn_prs_stress_res_part1_cuda(ta1%x_d, ta2%x_d, ta3%x_d, &
         wa1%x_d, wa2%x_d, wa3%x_d, f_x%x_d, f_y%x_d, f_z%x_d, &
         c_Xh%B_d, c_Xh%h1_d, rho%x_d, n)
#elif HAVE_OPENCL
!    call pnpn_prs_stress_res_part1_opencl(ta1%x_d, ta2%x_d, ta3%x_d, &
!         wa1%x_d, wa2%x_d, wa3%x_d, f_x%x_d, f_z%x_d, f_z%x_d, &
!         c_Xh%B_d, c_Xh%h1_d, rho%x_d, n)
#endif
```

**Review by njansson — COMMENTED** [review]
I would suggest to disable all missing backends, such that we can get this pass CI and into develop

**Comment by njansson** [comment]
> > I would suggest to disable all missing backends, such that we can get this pass CI and into develop
> 
> Or maybe throw an error from missing backends instead of doing nothing?

Indeed, changed

File: src/fluid/stress_formulation/bcknd/cpu/pnpn_res_stress_cpu.f90 [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+    call cmult(ta1%x, 2.0_rp, n)
+    call cmult(ta2%x, 2.0_rp, n)
+    call cmult(ta3%x, 2.0_rp, n)
+
+    ! S^T grad \mu
+    do e=1, nelv
+       call vdot3 (work1%x(:, :, :, e), &
+                   ta1%x(:, :, :, e), ta2%x(:, :, :, e), ta3%x(:, :, :,  e), &
+                   s11%x(:, :, :, e), s12%x(:, :, :, e), s13%x(:, :, :, e), lxyz)
```
Comment (timfelle) [inline src/fluid/stress_formulation/bcknd/cpu/pnpn_res_stress_cpu.f90:111]:
```suggestion
       call vdot3(work1%x(:, :, :, e), &
                  ta1%x(:, :, :, e), ta2%x(:, :, :, e), ta3%x(:, :, :, e), &
                  s11%x(:, :, :, e), s12%x(:, :, :, e), s13%x(:, :, :, e), lxyz)
```
You have a few extra spaces.

File: src/fluid/stress_formulation/bcknd/device/cuda/prs_stress_res_kernel.h [hunk 8]
```diff
@@ -0,0 +1,83 @@
+/*
+ Copyright (c) 2022, The Neko Authors
```
Comment (timfelle) [inline src/fluid/stress_formulation/bcknd/device/cuda/prs_stress_res_kernel.h:2]:
```suggestion
 Copyright (c) 2024, The Neko Authors
```

File: src/fluid/stress_formulation/pnpn_res_stress.f90 [hunk 9]
```diff
@@ -0,0 +1,157 @@
+! Copyright (c) 2022, The Neko Authors
```
Comment (timfelle) [inline src/fluid/stress_formulation/pnpn_res_stress.f90:1]:
```suggestion
! Copyright (c) 2024, The Neko Authors
```

**Review by timfelle — APPROVED** [review]

**Review by timfelle — APPROVED** [review]

**Pull request merged**